### DISCUSSION TITLE
How to do interleaved middleware

### DISCUSSION THREAD

**Discussion by daniel-white** [RobinTail/express-zod-api#1606] [Q&A] [answered]
With using the current middleware pattern, theres no way to interleave the request handling like traditional middleware - it expects the return value to become "options" for the handler, but that isn't sufficient for resources that need to be cleaned up or managed entirely within the middleware. I'd like these to be express-zod-api middlewares, since they is closer to the handlers.

For instance, this pattern is not possible:

```
function myMiddleware() {
  const client = createClient();
  try {
     // invoke handler with client
  } finally {
    client.close();
  }
} 
```

Examples: sql clients, opentelemetry span, etc.

**Comment by RobinTail** [comment]
I've never seen such a pattern in express.js , @daniel-white .
Could not even find "interleaved middleware".
Usually, middleware (the word) implies that it is in the middle, not around.

However, I understand your concern regarding managing resources.
Assuming that some libraries you use do not perform cleanup automatically upon instance deletion, I could offer you providing `options` to the Result Handler, where you could do a manual cleanup before responding to request.

Would that work for you?

**Reply by RobinTail** [reply]
Check this out, @daniel-white : 
https://github.com/RobinTail/express-zod-api/pull/1612

**Reply by RobinTail** [reply]
🚀 The feature released in version 17.4.0, @daniel-white

**Comment by RobinTail** [comment, marked as answer]
Update on this matter:
https://github.com/RobinTail/express-zod-api?tab=readme-ov-file#options
